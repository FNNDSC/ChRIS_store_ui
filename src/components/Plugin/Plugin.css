.plugin {
  /* display */
  min-height: calc(100vh - 66px);

  /* font */
  font-size: 1.2em;
}

.plugin, .plugin-container {
  /* display */
  display: flex;

  /* flex */
  flex-direction: column;
}

.plugin-container, .plugin-body {
  /* flex */
  flex: 1;
}

.plugin-header {
  /* display */
  padding: 1em;

  /* border */
  border-bottom-color: #bbbbbb;
  border-bottom: 1px solid var(--pf-black-400);
}

.plugin-category {
  /* font */
  text-transform: uppercase;
  font-weight: bold;
}

.plugin-name {
  /* color */
  color: #39a5dc;
  color: var(--pf-blue-300);
  
  /* font */
  font-size: 2.1em;
  font-weight: 300;
}

.plugin-name:focus, .plugin-name:hover {
  /* color */
  color: #0088ce;
  color: var(--pf-blue-400);

  /* font */
  text-decoration: none;
}

.plugin-modified {
  /* color */
  color: #bbbbbb;
  color: var(--pf-black-400);
}

.plugin-stats {
  /* display */
  display: block;
  margin-top: 1em;
}

.plugin-stats > .plugin-tag {
  /* color */
  color: #8b8d8f;
  color: var(--pf-black-500);
  
  /* display */
  display: inline-block;
  margin-right: 0.5em;
  padding: 0 0.25em;

  /* border */
  border-color: #8b8d8f;
  border: 1px solid var(--pf-black-500);
  border-radius: 4px;

  /* font */
  font-weight: 600;

  /* flex */
  flex: 1;
}

.plugin-author {
  /* color */
  color: #8b8d8f;
  color: var(--pf-black-500);
}

.plugin-author:hover, .plugin-author:focus {
  /* color */
  color: #72767b;
  color: var(--pf-black-600);
}
