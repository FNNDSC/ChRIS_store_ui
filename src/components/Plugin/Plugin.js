import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import StoreClient from '@fnndsc/chrisstoreapi';
import LoadingPlugin from './components/LoadingPlugin/LoadingPlugin';
import PluginBody from './components/PluginBody/PluginBody';
import RelativeDate from '../RelativeDate/RelativeDate';
import './Plugin.css';

const removeEmail = (author) => {
  if (author) return author.replace(/( ?\(.*\))/g, '');
  return undefined;
};

class Plugin extends Component {
  constructor(props) {
    super(props);

    this.mounted = false;

    const { pluginData } = props;
    this.state = { pluginData };

    this.fetchPluginData = this.fetchPluginData.bind(this);
  }

  componentWillMount() {
    this.mounted = true;
  }

  componentDidMount() {
    if (!this.state.pluginData) {
      this.fetchPluginData()
        .catch((err) => {
          console.error(err);
        });
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchPluginData() {
    const storeURL = process.env.REACT_APP_STORE_URL;
    const { plugin: pluginName } = this.props.match.params;
    const client = new StoreClient(storeURL);

    let pluginData;
    return new Promise(async (resolve, reject) => {
      try {
        pluginData = await client.getPlugin(pluginName);
      } catch (e) {
        return reject(e);
      }

      if (this.mounted) {
        this.setState({ pluginData });
      }
      return resolve(pluginData);
    });
  }

  render() {
    let plugin;
    let pluginURL;
    let authorURL;
    let data;

    const pluginDataProp = this.props.pluginData;
    if (pluginDataProp) {
      // define plugin name from props
      ({ plugin, pluginURL, authorURL } = pluginDataProp);

      data = pluginDataProp;
    } else {
      // define plugin name from url
      ({ plugin } = this.props.match.params);
      pluginURL = `/plugin/${plugin}`;

      const { pluginData } = this.state;
      data = pluginData || {};
    }

    // conditional rendering
    let container;
    if (Object.keys(data).length > 0) {
      const modificationDate = new RelativeDate(data.modification_date);
      const creationDate = new RelativeDate(data.creation_date);

      const author = removeEmail(data.authors);
      if (!authorURL) authorURL = `/author/${author}`;

      container = (
        <div className="plugin-container">
          <div className="plugin-header">
            <div className="row no-flex">
              <div className="plugin-category">Visualization</div>
              <Link
                href={pluginURL || '/'}
                to={pluginURL || '/'}
                className="plugin-name"
              >
                {data.name}
              </Link>
              {modificationDate.isValid() &&
                <div className="plugin-modified">
                  {`Last modified ${modificationDate.format()}`}
                </div>
              }
              <div className="plugin-stats">
                <div className="plugin-version plugin-tag">
                  {data &&
                    `v${data.version}`
                  }
                </div>
                <div className="plugin-created plugin-tag">
                  <Link
                    href={authorURL}
                    to={authorURL}
                    className="plugin-author"
                  >
                    {author}
                  </Link>
                  {creationDate.isValid() &&
                    ` created ${creationDate.format()}`
                  }
                </div>
                <div className="plugin-license plugin-tag">
                  {`${data.license} license`}
                </div>
              </div>
            </div>
          </div>
          <PluginBody pluginData={data} />
        </div>
      );
    } else {
      container = (
        <LoadingPlugin />
      );
    }

    return (
      <div className={`plugin ${this.props.className}`}>
        {container}
      </div>
    );
  }
}

Plugin.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      plugin: PropTypes.string,
    }),
  }),
  pluginData: PropTypes.shape({
    plugin: PropTypes.string,
    pluginURL: PropTypes.string,
    authorURL: PropTypes.string,
    title: PropTypes.string,
    id: PropTypes.number,
    description: PropTypes.string,
    dock_image: PropTypes.string,
    modification_date: PropTypes.string,
    creation_date: PropTypes.string,
    authors: PropTypes.string,
    version: PropTypes.string,
  }),
  className: PropTypes.string,
};

Plugin.defaultProps = {
  match: {
    params: {
      plugin: undefined,
    },
  },
  pluginData: null,
  className: '',
};

export default Plugin;
