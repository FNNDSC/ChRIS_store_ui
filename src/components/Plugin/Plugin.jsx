import React, { useEffect, useState } from 'react';
import { Badge, Grid, GridItem, Split, SplitItem, Button } from '@patternfly/react-core';
import { StarIcon } from '@patternfly/react-icons';
import PropTypes from 'prop-types';
import Client, { Plugin } from '@fnndsc/chrisstoreapi';

import LoadingPlugin from './components/LoadingPlugin/LoadingPlugin';
import PluginBody from './components/PluginBody/PluginBody';
import RelativeDate from '../RelativeDate/RelativeDate';
import ChrisStore from '../../store/ChrisStore';
import PluginImg from '../../assets/img/brainy-pointer.png';
import NotFound from '../NotFound/NotFound';
import ErrorNotification from '../Notification';
import HttpApiCallError from '../../errors/HttpApiCallError';

import './Plugin.css';

/**
 * View a plugin by plugin ID.
 */
const PluginView = (props) => {
const [state, setState] = useState({
  pluginData: undefined,
  star: undefined,
  loading: true,
  errors: [],
})

    const { match, store } = props
    const { loading, pluginData: plugin, errors, star } = state;

    const storeURL = process.env.REACT_APP_STORE_URL;
    const auth = { token: store.get('authToken') };
    const client = new Client(storeURL, auth);

const showNotifications = (error) => {
    setState((prev) => ({
      errors: [...prev.errors, error]
    }));
  }

const isFavorite = () => star !== undefined;

const isLoggedIn = () => store ? store.get('isLoggedIn') : false;

const favPlugin = async () => {
    const { pluginData } = state;
    // Early state change for instant visual feedback
    pluginData.stars += 1;
    setState({ star: {}, pluginData });

    try {
      const createdStar = await client.createPluginStar({ plugin_name: pluginData.name });
      setState({ star: createdStar.data });
    } catch (error) {
      showNotifications(new HttpApiCallError(error));
      pluginData.stars -= 1;
      setState({ star: undefined, pluginData });
    }
  }

const unfavPlugin = async () => {
    const { pluginData, star: previousStarState } = state;

    // Early state change for instant visual feedback
    pluginData.stars -= 1;
    setState({ star: undefined, pluginData });

    try {
      await (
        await client.getPluginStar(previousStarState.id)
      ).delete();
    } catch (error) {
      pluginData.stars += 1;
      setState({ star: previousStarState, pluginData });
      showNotifications(new HttpApiCallError(error));
    }
  }

  const onStarClicked = () => {
    if (isLoggedIn()) {
      if (isFavorite())
        unfavPlugin();
      else
        favPlugin();
    }
    else
      showNotifications(new Error('Login required to favorite this plugin.'))
  }

// const renderStar = () => {
//     let name;
//     let className;

//     if (isLoggedIn()) {
//       className = isFavorite() ? 'plugin-star-favorite' : 'plugin-star';
//       name = isFavorite() ? 'star' : 'star-o';
//     } else {
//       className = 'plugin-star-disabled';
//       name = 'star-o';
//     }
//     return <StarIcon name={name} className={className} onClick={() => onStarClicked()} />;
//   }

  /**
   * Fetch a plugin by ID
   * @param {string} pluginId 
   * @returns {Promise} Plugin
   */
const fetchPlugin = async (pluginId) => client.getPlugin(parseInt(pluginId, 10));

  /**
   * Fetch all versions of a plugin by name.
   * @param {string} name Plugin name
   * @returns Promise => void
   */
const fetchPluginVersions = async (name) => {
    const versions = await client.getPlugins({ limit: 10e6, name_exact: name });
    const firstplg = await client.getPlugin(parseInt(versions.data[0].id, 10));
    return [
      { ...versions.data[0], url: firstplg.url },
      ...versions.data.slice(1)
    ]
  }

const fetchIsPluginStarred = async ({ name }) => {
    const response = await client.getPluginStars({ plugin_name: name });
    if (response.data.length > 0)
      return response.data[0];
    return undefined;
  }

      /**
   * Fetch a plugin by ID, from URL params.
   * Then fetch other plugins which have the same name as versions.
   * Set stars if user is logged in.
   */
useEffect(() => {
  const { pluginId } = match.params;
  const fetchData = async() => {
  try {
    const fetchedPlugin = await fetchPlugin(pluginId);
    const versions = await fetchPluginVersions(fetchedPlugin.data.name);

    // eslint-disable-next-line no-shadow
    let star;
    if (isLoggedIn())
      star = await fetchIsPluginStarred(plugin.data);

    setState({
      loading: false,
      pluginData: {
        ...plugin.data,
        url: plugin.url,
        versions
      },
      star,
    });
  } catch (error) {
    setState((prev) => ({
      loading: false,
      errors: [...prev.errors, error]
    }));
  }
}
fetchData()
})
    if (!loading && !plugin)
      return <NotFound />

    let container;
    if (plugin) {
      container = (
        <article>
          <section>
            <Grid hasGutter>
              <GridItem style={{ marginRight: '2em' }} lg={2} xs={12}>
                <img
                  className="plugin-icon"
                  src={PluginImg}
                  alt="Plugin icon"
                />
              </GridItem>

              <GridItem lg={10} xs={12}>
                <Grid>
                  <GridItem lg={10} xs={12}>
                    <h3 className="plugin-name">{plugin.name} <Badge>{plugin.category}</Badge></h3>
                    <h2 className="plugin-title">{plugin.title}</h2>
                  </GridItem>

                  <GridItem lg={2} xs={12} className="plugin-stats">
                    <Split>
                      <SplitItem isFilled />
                      <SplitItem>
                        {
                          !isFavorite() ?
                            <Button onClick={() => onStarClicked()}>
                              Favorite <Badge isRead><StarIcon /> {plugin.stars}</Badge>
                            </Button>
                            :
                            <Button variant="secondary" onClick={() => onStarClicked()}>
                              Unfavorite <Badge><StarIcon /> {plugin.stars}</Badge>
                            </Button>
                        }
                      </SplitItem>
                    </Split>
                  </GridItem>

                  <GridItem>
                    <p>{plugin.description}</p>
                    <p style={{ color: "gray" }}>
                      {
                        RelativeDate.isValid(plugin.modification_date) ?
                          `Updated ${new RelativeDate(plugin.modification_date).format()}`
                          :
                          `Created ${new RelativeDate(plugin.creation_date).format()}`
                      }
                    </p>
                  </GridItem>
                </Grid>
              </GridItem>
            </Grid>
          </section>

          <section>
            <PluginBody pluginData={plugin} />
          </section>
        </article>
      );
    } else {
      container = (
        <article>
          <LoadingPlugin />
        </article>
      );
    }

    return (
      <>
        {
          errors.map((message, index) => (
            <ErrorNotification
              key={`notif-${message}`}
              title={message}
              position='top-right'
              variant='danger'
              closeable
              onClose={() => {
                errors.splice(index)
                setState({ errors })
              }}
            />
          ))
        }
        <div className="plugin">
          {container}
        </div>
      </>

    );
  }


Plugin.propTypes = {
  store: PropTypes.objectOf(PropTypes.object),
  match: PropTypes.shape({
    params: PropTypes.shape({
      plugin: PropTypes.string,
    })
  })
};

Plugin.defaultProps = {
  store: new Map(),
  match: {
    params: {
      plugin: undefined,
    }
  }
};

export default ChrisStore.withStore(PluginView);
