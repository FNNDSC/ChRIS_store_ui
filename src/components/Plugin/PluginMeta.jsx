import React, { Component } from 'react';
import { Badge, Grid, GridItem, Split, SplitItem, Button } from '@patternfly/react-core';
import { StarIcon } from '@patternfly/react-icons';
import PropTypes from 'prop-types';
import Client, { PluginMeta } from '@fnndsc/chrisstoreapi';

import LoadingPlugin from './components/LoadingPlugin/LoadingPlugin';
import PluginBody from './components/PluginBody/PluginBody';
import RelativeDate from '../RelativeDate/RelativeDate';
import ChrisStore from '../../store/ChrisStore';
import PluginImg from '../../assets/img/brainy-pointer.png';
import NotFound from '../NotFound/NotFound';
import ErrorNotification from '../Notification';
import HttpApiCallError from '../../errors/HttpApiCallError';

import './Plugin.css';

/**
 * View a plugin meta by plugin name.
 * 
 * @todo 
 * Make this view visually different
 * from the plugin view by ID.
 */
export class PluginMetaView extends Component {
  constructor(props) {
    super(props);

    this.state = {
      pluginData: undefined,
      star: undefined,
      loading: true,
      errors: [],
    };

    const storeURL = process.env.REACT_APP_STORE_URL;
    const auth = { token: props.store.get('authToken') };
    this.client = new Client(storeURL, auth);
  }

  /**
   * Fetch a plugin meta by name, from URL params.
   * Then fetch all versions of that plugin.
   * Set stars if user is logged in.
   */
  async componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    const { pluginName } = this.props.match.params;
    try {
      const pluginMeta = await this.fetchPluginMeta(pluginName);
      const versions = await this.fetchPluginVersions(pluginMeta);

      let star;
      if (this.isLoggedIn())
        star = await this.fetchIsPluginStarred(pluginMeta.data);

      this.setState({ 
        loading: false,
        star,
        pluginData: {
          ...pluginMeta.data,
          versions
        }
      });
    } catch (error) {
      this.setState((prev) => ({ 
        loading: false, 
        errors: [ ...prev.errors, error ] 
      }));
    }
  }
  
  showNotifications = (error) => {
    this.setState((prev) => ({
      errors: [ ...prev.errors, error ] 
    }));
  }

  // eslint-disable-next-line react/destructuring-assignment
  isFavorite = () => this.state.star !== undefined;

  // eslint-disable-next-line react/destructuring-assignment
  isLoggedIn = () => this.props.store ? this.props.store.get('isLoggedIn') : false;

  onStarClicked = () => {
    if (this.isLoggedIn()) {
      if (this.isFavorite()) 
        this.unfavPlugin();
      else 
        this.favPlugin();
    }
    else
      this.showNotifications(new Error('Login required to favorite this plugin.'))
  }

  favPlugin = async () => {
    const { pluginData } = this.state;

    // Early state change for instant visual feedback
    pluginData.stars += 1;
    this.setState({ star: {}, pluginData });

    try {
      const star = await this.client.createPluginStar({ plugin_name: pluginData.name });
      this.setState({ star: star.data });
    } catch (error) {
      this.showNotifications(new HttpApiCallError(error));
      pluginData.stars -= 1;
      this.setState({ star: undefined, pluginData });
    }
  }

  unfavPlugin = async () => {
    const { pluginData, star: previousStarState } = this.state;

    // Early state change for instant visual feedback
    pluginData.stars -= 1;
    this.setState({ star: undefined, pluginData });

    try {
      await (
        await this.client.getPluginStar(previousStarState.id)
      ).delete();
    } catch (error) {
      pluginData.stars += 1;
      this.setState({ star: previousStarState, pluginData });
      this.showNotifications(new HttpApiCallError(error));
    }
  }

  /**
   * Fetch a plugin meta by plugin name.
   * @param {string} pluginName 
   * @returns {Promise<PluginMeta>} PluginMeta
   */
  async fetchPluginMeta(pluginName) {
    const metas = await this.client.getPluginMetas({ name_exact: pluginName, limit: 1 });
    return metas.getItems().shift();
  }

  /**
   * Fetch all versions of a plugin.
   * @param {PluginMeta} pluginMeta 
   * @returns {Promise<any[]>} Versions of the plugin
   */
  // eslint-disable-next-line class-methods-use-this
  async fetchPluginVersions(pluginMeta) {
    const versions = (await pluginMeta.getPlugins()).getItems();
    return versions.map(({ data, url }) => ({ ...data, url }));
  }

  async fetchIsPluginStarred({ name }) {
    const response = await this.client.getPluginStars({ plugin_name: name });
    if (response.data.length > 0)
      return response.data[0];
    return undefined;
  }

  render() {
    const { loading, pluginData: plugin, errors } = this.state;

    if (!loading && !plugin)
      return <NotFound/>

    let container;
    if (plugin) {
      container = (
        <article>
          <section>
            <Grid hasGutter>
              <GridItem style={{ marginRight: '2em' }} lg={2} xs={12}>
                <img
                  className="plugin-icon"
                  src={PluginImg}
                  alt="Plugin icon"
                />
              </GridItem>

              <GridItem lg={10} xs={12}>
                <Grid>
                  <GridItem lg={10} xs={12}>
                    <h3 className="plugin-name">{plugin.name} <Badge>{plugin.category}</Badge></h3>
                    <h2 className="plugin-title">{plugin.title}</h2>
                  </GridItem>

                  <GridItem lg={2} xs={12} className="plugin-stats">
                    <Split>
                      <SplitItem isFilled />
                      <SplitItem>
                        {
                          !this.isFavorite() ? 
                            <Button onClick={this.onStarClicked}>
                              Favorite <Badge isRead><StarIcon /> {plugin.stars}</Badge>
                            </Button>
                          : 
                            <Button variant="secondary" onClick={this.onStarClicked}>
                              Unfavorite <Badge><StarIcon /> {plugin.stars}</Badge>
                            </Button>
                        }
                      </SplitItem>
                    </Split>
                  </GridItem>

                  <GridItem>
                    <p>{plugin.description}</p>
                    <p style={{ color: "gray" }}>
                      { 
                        RelativeDate.isValid(plugin.modification_date) ?
                          `Updated ${new RelativeDate(plugin.modification_date).format()}`
                        : 
                          `Created ${new RelativeDate(plugin.creation_date).format()}`
                      }
                    </p>
                  </GridItem>
                </Grid>
              </GridItem>
            </Grid>
          </section>

          <section>
            <PluginBody pluginData={plugin} />
          </section>
        </article>
      );
    } else {
      container = (
        <article>
          <LoadingPlugin />
        </article>
      );
    }

    return (
      <>
        {
          errors.map((error, index) => (
            <ErrorNotification
              key={`notif-${error.message}`}
              title="Error"
              message={error.message}
              position='topRight'
              variant='danger'
              closeable
              onClose={() => {
                errors.splice(index)
                this.setState({ errors })
              }}
            />
          ))
        }
        <div className="plugin">
          {container}
        </div>
      </>

    );
  }
}

PluginMeta.propTypes = {
  store: PropTypes.objectOf(PropTypes.object),
  match: PropTypes.shape({
    params: PropTypes.shape({
      plugin: PropTypes.string,
    })
  })
};

PluginMeta.defaultProps = {
  store: new Map(),
  match: {
    params: {
      plugin: undefined,
    }
  }
};

export default ChrisStore.withStore(PluginMetaView);
