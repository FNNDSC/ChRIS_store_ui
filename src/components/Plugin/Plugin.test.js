import React from 'react';
import { shallow } from 'enzyme';
import Plugin from './Plugin';

// define mock for @fnndsc/chrisstoreapi module
jest.mock('@fnndsc/chrisstoreapi', () => require.requireActual('../__mocks__/chrisstoreapi').default);

describe('Plugin', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Plugin />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a plugin div', () => {
    expect(wrapper.find('div.plugin')).toHaveLength(1);
  });

  it('should render plugin div based on className prop', () => {
    expect(wrapper.find('div.plugin.testClass')).toHaveLength(0);
    wrapper.setProps({ className: 'testClass' });
    expect(wrapper.find('div.plugin.testClass')).toHaveLength(1);
  });

  /* ============================== */
  /* =========== METHODS ========== */
  /* ============================== */

  it('sets mounted to true on after mounting', () => {
    expect(Plugin.prototype.mounted).toBeFalsy();
    Plugin.prototype.componentWillMount();
    expect(Plugin.prototype.mounted).toBeTruthy();
  });

  it('sets mounted to false after componentWillUnmount is called', () => {
    const instance = wrapper.instance();
    expect(instance.mounted).toBeTruthy();
    instance.componentWillUnmount();
    expect(instance.mounted).toBeFalsy();
  });

  /* ============================== */
  /* ==== FETCH PLUGIN DATA FN ==== */
  /* ============================== */

  it('should have fetchPluginData function', () => {
    const { fetchPluginData } = wrapper.instance();
    expect(fetchPluginData).toBeDefined();
  });

  it('should be able to fetch plugin data', () => {
    const { fetchPluginData } = wrapper.instance();

    wrapper.setProps({
      match: {
        params: {
          plugin: 'dircopy',
        },
      },
    });

    return fetchPluginData().then((pluginData) => {
      expect(pluginData).toBeDefined();
      expect(pluginData.name).toEqual('testName1');
    });
  });

  const propsData = {
    modification_date: '3/14/15',
    creation_date: '9/26/53',
    version: '0.1',
    authors: 'testAuthor (user@domain.com)',
  };

  it('should set pluginData state to pluginData props if passed', () => {
    const wrapperWithoutProps = shallow(<Plugin />);
    expect(wrapperWithoutProps.state('pluginData')).toBeNull();
    const wrapperWithProps = shallow(<Plugin pluginData={propsData} />);
    expect(wrapperWithProps.state('pluginData')).toEqual(propsData);
  });
});

/* ============================== */
/* ===== WITHOUT PLUGIN DATA ==== */
/* ============================== */

describe('Plugin without data', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Plugin />);
  });

  it('should render LoadingPlugin component', () => {
    expect(wrapper.find('LoadingPlugin')).toHaveLength(1);
  });
});

/* ============================== */
/* ====== WITH PLUGIN DATA ====== */
/* ============================== */

const sampleData = {
  pluginData: {
    modification_date: '3/14/15',
    version: '0.1',
    authors: 'testAuthor (user@domain.com)',
  },
};

describe('Plugin with data', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Plugin />);
    wrapper.setState(sampleData);
  });

  it('should render a plugin-name Link', () => {
    expect(wrapper.find('Link.plugin-name')).toHaveLength(1);
  });

  it('should render plugin name', () => {
    wrapper.setProps({
      match: {
        params: {
          plugin: 'testName1',
        },
      },
    });

    expect(wrapper
      .find('.plugin-name')
      .childAt(0)
      .text())
      .toEqual('testName1');
  });

  it('plugin-name Link should render correct "to" and "href" props', () => {
    wrapper.setProps({
      match: {
        params: {
          plugin: 'testName',
        },
      },
    });

    const link = wrapper.find('Link.plugin-name');
    const whatItShould = '/plugin/testName';
    expect(link.prop('to')).toEqual(whatItShould);
    expect(link.prop('href')).toEqual(whatItShould);
  });

  it('should render plugin-modified div', () => {
    expect(wrapper.find('div.plugin-modified')).toHaveLength(1);
  });

  it('should render plugin-modified div if valid date is provided', () => {
    Date.now = jest.fn(() => 1530814238992);
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.modification_date = '2018-06-19T15:29:11.349272Z';
    wrapper.setState(changedData);

    expect(wrapper
      .find('div.plugin-modified')
      .text())
      .toEqual('Last modified 16 days ago');
  });

  it('should not render plugin-modified if invalid date is provided', () => {
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.modification_date = 'invalid date';

    wrapper.setState(changedData);

    expect(wrapper.find('div.plugin-modified')).toHaveLength(0);
  });

  it('should render plugin-stats div', () => {
    expect(wrapper.find('div.plugin-stats')).toHaveLength(1);
  });

  it('should render plugin-version div inside plugin-stats', () => {
    expect(wrapper
      .find('div.plugin-stats')
      .find('div.plugin-version'))
      .toHaveLength(1);
  });

  it('should render correct plugin version inside plugin-version div', () => {
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.version = 'testVersion';
    wrapper.setState(changedData);

    expect(wrapper
      .find('div.plugin-version')
      .text())
      .toEqual('vtestVersion');
  });

  it('should render plugin-created div inside plugin-stats', () => {
    expect(wrapper
      .find('div.plugin-stats')
      .find('div.plugin-created'))
      .toHaveLength(1);
  });

  it('should render Link inside plugin-created', () => {
    expect(wrapper
      .find('div.plugin-created')
      .find('Link.plugin-author'))
      .toHaveLength(1);
  });

  it('plugin-author Link should render correct "to" and "href" props', () => {
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.authors = 'testAuthor (user@domain.com)';
    wrapper.setState(changedData);

    const link = wrapper.find('Link.plugin-author');
    const whatItShould = '/author/testAuthor';
    expect(link.prop('to')).toEqual(whatItShould);
    expect(link.prop('href')).toEqual(whatItShould);
  });

  it('plugin-author Link should render correct text', () => {
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.authors = 'testAuthor (user@domain.com)';
    wrapper.setState(changedData);

    expect(wrapper
      .find('Link.plugin-author')
      .childAt(0)
      .text())
      .toEqual('testAuthor');
  });

  it('plugin-author Link should render "to" and "href" props based on pluginData.authorURL prop', () => {
    const pluginData = {
      authorURL: '/test/url',
    };
    wrapper.setProps({ pluginData });

    const link = wrapper.find('Link.plugin-author');
    expect(link.prop('to')).toEqual('/test/url');
    expect(link.prop('href')).toEqual('/test/url');
  });

  const getPluginCreatedText = () => {
    const pluginCreationDiv = wrapper.find('div.plugin-created');
    const receivedText = pluginCreationDiv.text();
    // remove html elements
    return receivedText.replace(/(<.*>)/g, '');
  };

  it('should render the value of creationDate if it is a valid date', () => {
    Date.now = jest.fn(() => 1530814238992);
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.creation_date = '2018-06-19T15:29:11.349272Z';
    wrapper.setState(changedData);
    expect(getPluginCreatedText()).toEqual(' created 16 days ago');
  });

  it('should not render the value of creationDate if it is not a valid date', () => {
    const changedData = Object.assign({}, sampleData);
    changedData.pluginData.creation_date = 'invalid date';
    wrapper.setState(changedData);
    expect(getPluginCreatedText()).toEqual('');
  });

  it('should render PluginBody component inside plugin-container', () => {
    expect(wrapper
      .find('div.plugin-container')
      .find('PluginBody'))
      .toHaveLength(1);
  });
});
