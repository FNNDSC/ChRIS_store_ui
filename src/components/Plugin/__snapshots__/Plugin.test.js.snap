// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Plugin
    className=""
    match={
        Object {
            "params": Object {
              "plugin": undefined,
            },
          }
    }
    pluginData={null}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <LoadingPlugin />,
      "className": "plugin ",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <LoadingPlugin />,
        "className": "plugin ",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
